### PR TITLE
CLI: Add Server template generator

### PR DESCRIPTION
**Pull request by jonspalmer**
Issue: #12077

## What I did

Added Cli generator support for Server

## Open questions

It would be helpful to the user if the generator added a few stanzas to `.storybook/preview.js` to template the configuration required to point to the server backend. However, I'm a little lost as to have to arrange that. It seems like the template copying has mostly been eliminated and preview in particular gets written explicitly here: https://github.com/storybookjs/storybook/blob/next/lib/cli/src/generators/configure.ts#L18

Not sure if the intent would be to add more cases to that logic or somehow generalize the preview handling to make it simpler for apps to provide their own tweaks.

### CODE REVIEW

File: lib/cli/src/generators/SERVER/index.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import fse from 'fs-extra';
+import { baseGenerator, Generator } from '../baseGenerator';
+
+const generator: Generator = async (packageManager, npmOptions, options) => {
+  const prefix = fse.existsSync('./src') ? '../src' : '../stories';
+  const stories = [`${prefix}/**/*.stories.json`];
```
Comment (phated) [inline lib/cli/src/generators/SERVER/index.ts:6]: People could have stories in any location, correct? Maybe we should just prescribe one like we do in other generators.

File: lib/cli/src/generators/SERVER/index.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+  baseGenerator(packageManager, npmOptions, options, 'server', {
+    addComponents: false,
+    configureOptions: { stories },
```
Comment (shilman) [inline lib/cli/src/generators/SERVER/index.ts:10]: Doesn't seem like this is used?

**Review by shilman — CHANGES_REQUESTED** [review]

File: lib/cli/src/generators/SERVER/index.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { baseGenerator, Generator } from '../baseGenerator';
+
+const generator: Generator = async (packageManager, npmOptions, options) => {
+  const stories = ['../stories/**/*.stories.json'];
+
+  baseGenerator(packageManager, npmOptions, options, 'server', {
+    addComponents: false,
+    extraMain: { stories },
```
Comment (jonspalmer) [inline lib/cli/src/generators/SERVER/index.ts:8]: updated to leverage the recently added `extraMain` params.

File: lib/cli/src/generators/SERVER/index.ts (hunk 1 above)
Comment (jonspalmer) [reply lib/cli/src/generators/SERVER/index.ts]: Simplified this.

File: lib/cli/src/generators/SERVER/index.ts (hunk 2 above)
Comment (jonspalmer) [reply lib/cli/src/generators/SERVER/index.ts]: reworked it a little

**Comment by shilman** [comment]
In all of our other project templates, `yarn storybook` should work out of the box. Currently that's not the case in this PR.

Would it be possible to:
1) generate some simple stories
2) generate a hello world style server (using something built in to node?)

That way everything's hooked up and the user can just remove stuff to get it working with their own server. What do you think?

**Comment by jonspalmer** [comment]
@shilman I think we're in decent shape here. Did you solidify the netlify app that backs these generated samples?

**Comment by shilman** [comment]
@jonspalmer I think it's in good shape

**Review by shilman — APPROVED** [review]
LGTM!

**Comment by yannbf** [comment]
@jonspalmer @shilman the url http://storybook-server-demo.netlify.app/api resolves into a 404. Did we miss something?

**Comment by shilman** [comment]
@yannbf thanks! i added minimal docs and a redirect to the docs. it never actually called that URL, so it didn't matter in practice that it was a 404. but the redirect removes more opportunities for confusion.

**Pull request merged**